src/Microsoft.ML.Tokenizers/Model/SentencePieceUnigramModel.cs

```diff
@@ -0,0 +1,1399 @@
+                    _minScore = Math.Min(_minScore, score);
+                    _maxScore = Math.Max(_maxScore, score);
+                }
+                else if (modelProto.Pieces[i].Type == ModelProto.Types.SentencePiece.Types.Type.Byte)
+                {
+                    MaxByteId = i;
+                }
+                else if (modelProto.Pieces[i].Type == ModelProto.Types.SentencePiece.Types.Type.Unknown)
+                {
+                    // Ensure the unknown token is cached
+                    _vocabReverse[i] = (modelProto.Pieces[i].Piece, modelProto.Pieces[i].Score, ModelProto.Types.SentencePiece.Types.Type.Unknown);
+                }
+            }
+
+            ByteCodeToIdOffset = _vocab.TryGetValue("<0x00>", out int id) ? id : MaxByteId;
+            OneByteUtf8EncodingMaxId = ByteCodeToIdOffset + 0x7F; // 0x7F is the maximum value of the one byte UTF-8 character.
+            MaxIdByteFallbackId = ByteCodeToIdOffset + 0xFF; // from <0x00> to <0xFF>.
+
+            _trie = new DoubleArrayTrie(_vocab);
+
+            _vocabReverse[BeginningOfSentenceId] = (BeginningOfSentenceToken, 0f, 0);
+            _vocabReverse[EndOfSentenceId] = (EndOfSentenceToken, 0f, 0);
+        }
+
+        public override IReadOnlyDictionary<string, int> Vocabulary => new ReadOnlyDictionary<string, int>(_vocab);
+
+        public int MaxIdByteFallbackId { get; }
+
+        public override IReadOnlyList<EncodedToken> EncodeToTokens(string? text, ReadOnlySpan<char> textSpan, out string? normalizedText, bool addBeginningOfSentence, bool addEndOfSentence, bool considerNormalization)
+        {
+            ReadOnlySpan<char> textToEncode = string.IsNullOrEmpty(text) ? textSpan : text.AsSpan();
+            if (textToEncode.IsEmpty)
+            {
+                normalizedText = string.Empty;
+                return Array.Empty<EncodedToken>();
+            }
+
+            List<EncodedToken> tokens = new();
+
+            // Rent a buffer that approximately enough to hold the Utf8 encoded bytes, the normalization of the encoded buffer, and some extra memory to for encoding results.
+            int[] buffer = ArrayPool<int>.Shared.Rent(textToEncode.Length * 3);
+
+            // Hold the Utf16 normalized string.
+            char[] normalizedString = ArrayPool<char>.Shared.Rent(textToEncode.Length + 2);
+
+            if (SpecialTokensRegex is not null)
+            {
+                EncodeToTokensWithSpecialTokens(textToEncode, addBeginningOfSentence, addEndOfSentence, considerNormalization, tokens, buffer, ref normalizedString, out normalizedText);
+            }
+            else
+            {
+                EncodeToTokensWithoutSpecialTokens(textToEncode, addBeginningOfSentence, addEndOfSentence, considerNormalization, tokens, buffer, ref normalizedString, out normalizedText);
+            }
+
+            ArrayPool<char>.Shared.Return(normalizedString);
+            ArrayPool<int>.Shared.Return(buffer);
```

Comment:
Should these be in try/finally?

Reply:
Not necessary. If something goes wrong the rented array will be garbage collected. nothing hurt.